### PR TITLE
Ray so

### PR DESCRIPTION
**Pull request by gt-codes**
## Description

This extension allows you to create a beautiful code snippet right from Raycast

## Type of change

- New extension

## Screencast

[Ray.so Extension Screencast](https://loom.com/share/10d9d2485be3417fa1a5dbbad29f3c5f)

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I checked that extension's icon has size of `512x512`

### CODE REVIEW

File: extensions/ray-so/src/create-a-snippet.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        ))}
+      </Form.Dropdown>
+      <Form.Dropdown
+        id="background"
+        title="Background"
+        storeValue
+        onChange={(background) => setCode({ ...code, background })}
+      >
+        <Form.Dropdown.Item value="true" title="True" />
+        <Form.Dropdown.Item value="false" title="False" />
```
Comment (pernielsentikaer) [inline extensions/ray-so/src/create-a-snippet.tsx:73]:
```suggestion
        <Form.Dropdown.Item value="true" title="Yes" />
        <Form.Dropdown.Item value="false" title="No" />
```

File: extensions/ray-so/src/create-a-snippet.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      </Form.Dropdown>
+      <Form.Dropdown id="darkMode" title="Dark Mode" storeValue onChange={(darkMode) => setCode({ ...code, darkMode })}>
+        <Form.Dropdown.Item value="true" title="True" />
+        <Form.Dropdown.Item value="false" title="False" />
```
Comment (pernielsentikaer) [inline extensions/ray-so/src/create-a-snippet.tsx:77]:
```suggestion
        <Form.Dropdown.Item value="true" title="Yes" />
        <Form.Dropdown.Item value="false" title="No" />
```

File: extensions/ray-so/src/create-a-snippet.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        {colors.map((el, idx) => (
+          <Form.Dropdown.Item
+            key={idx}
+            icon={{ source: `${el}.png` }}
+            value={el}
+            title={el.charAt(0).toUpperCase() + el.substring(1).toLowerCase()}
+          />
+        ))}
+      </Form.Dropdown>
+      <Form.Dropdown id="langauge" title="Language" storeValue onChange={(language) => setCode({ ...code, language })}>
```
Comment (pernielsentikaer) [inline extensions/ray-so/src/create-a-snippet.tsx:61]:
```suggestion
      <Form.Dropdown id="language" title="Language" storeValue onChange={(language) => setCode({ ...code, language })}>
```
Just a typo

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @Garrett1Tolbert 👋 

That's a beautiful extension and only minor NITs before we're ready to merge it, very good job, well done 💪 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

File: extensions/ray-so/src/create-a-snippet.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import { Form, ActionPanel, OpenInBrowserAction } from "@raycast/api";
+import { colors, languages } from "./constants";
+import { useState } from "react";
```
Comment (pernielsentikaer) [inline extensions/ray-so/src/create-a-snippet.tsx:3]:
```suggestion
import { useState } from "react";
import open from "open";
```

File: extensions/ray-so/src/create-a-snippet.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
+import { Form, ActionPanel, OpenInBrowserAction } from "@raycast/api";
```
Comment (pernielsentikaer) [inline extensions/ray-so/src/create-a-snippet.tsx:1]:
```suggestion
import { Form, ActionPanel, SubmitFormAction } from "@raycast/api";
```

File: extensions/ray-so/src/create-a-snippet.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  snippet: "",
+  color: "candy",
+};
+
+export default function CreateSnippet() {
+  const [code, setCode] = useState<Values>(defaultSnippet);
+  const url = `https://ray.so/?colors=${code.color}&background=${code.background}&darkMode=${code.darkMode}&padding=${
+    code.padding
+  }&title=${code.title || "Untitled 1"}&code=${Buffer.from(code.snippet).toString("base64")}&language=${code.language}`;
+
```
Comment (pernielsentikaer) [inline extensions/ray-so/src/create-a-snippet.tsx:29]:
```suggestion

  function handleSubmit() {
    open(url);
  }
```

File: extensions/ray-so/src/create-a-snippet.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  return (
+    <Form
+      actions={
+        <ActionPanel>
+          <OpenInBrowserAction title="Create Snippet"  url={url} />
```
Comment (pernielsentikaer) [inline extensions/ray-so/src/create-a-snippet.tsx:34]:
```suggestion
          <SubmitFormAction title="Create Snippet" onSubmit={handleSubmit} />
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @Garrett1Tolbert  👋 

It looks a lot better already, we're close to merging it!

A thing I really think could be enhanced is the possibility to use enter as newline instead of submitting the form. See suggestions in code on how to archive this.

(requires npm open). Hopefully I didn't forget anything 😄 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Review by pernielsentikaer — APPROVED** [review]
Hi @Garrett1Tolbert 👋 

I think we're ready to get this 💎 merged, thanks for the contribution 💪

File: extensions/ray-so/package.json [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  "name": "ray-so",
+  "title": "Ray.so",
+  "description": "Create beautiful images of your code",
+  "icon": "icon.png",
+  "author": "Garrett",
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "create-a-snippet",
+      "title": "Create Beautiful Images of Code",
```
Comment (PitNikola) [inline extensions/ray-so/package.json:12]:
```suggestion
      "title": "Create Image from Code",
```

I would suggest naming for similar [script command](https://github.com/raycast/script-commands/blob/master/commands/developer-utils/create-image-from-code.sh)

File: extensions/ray-so/package.json [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+      "subtitle": "Raycast",
```
Comment (PitNikola) [inline extensions/ray-so/package.json:13]:
```suggestion
      "subtitle": "Ray.so",
```

File: extensions/ray-so/package.json [hunk 10]
```diff
@@ -0,0 +1,36 @@
+{
+  "$schema": "https://www.raycast.com/schemas/extension.json",
+  "name": "ray-so",
+  "title": "Ray.so",
+  "description": "Create beautiful images of your code",
```
Comment (PitNikola) [inline extensions/ray-so/package.json:5]:
```suggestion
  "description": "Create beautiful images of your code using ray.so",
```

**Review by PitNikola — CHANGES_REQUESTED** [review]
Hey @Garrett1Tolbert, nice extension, thanks for creating it!

Few suggestions:
- Let's not use the default extension icon, it's not really for publishing in the store. Since you're connecting with our service, I went ahead and created an icon for you, feel free to just use it.
ray-so-icon
- Shall we add some validation for when nothing is entered? I would suggest showing a toast saying that text area is empty

Also I'm experiencing some issues with `open` code not actually working, will investigate and get back to you. In the meantime you can address other comments.

Thanks!

**Comment by Garrett1Tolbert** [comment]
@PitNikola Not sure why the check isn't passing but one quick thing – I made the icon have a transparent background. Does it look good on your end?

**Comment by pernielsentikaer** [comment]
Hi @Garrett1Tolbert 👋 

It seems like it has nothing to do with the logo, if checking the log it says that npm should be updated — can you handle that? We've added new validation since you last push 😄 

```
💥 Error: wrong 'lockfileVersion' in package-lock.json. Found "lockfileVersion: 1", expected: '2'. Update your 'npm' to v7.x
Error: 💥 Error: wrong 'lockfileVersion' in package-lock.json. Found "lockfileVersion: 1", expected: '2'. Update your 'npm' to v7.x 
Error: Process completed with exit code 1.
```

**Comment by pernielsentikaer** [comment]
Hi @Garrett1Tolbert 👋 

I think the toast with an error message as @PitNikola mentioned could be the last thing, right now It's possible to send the form without writing anything which is not the best UX

**Comment by PitNikola** [comment]
Sorry, I just realised that I exported the icon with the background. Here's transparent one:

ray-so-icon

Now as for the issue that I had – I found the repro. Seems like `open` doesn't work when you have space in URL. This got broken on Big Sur when Apple decided to change its behavior for some reason. Could you please confirm that it happens whenever you have title with a space (e.g. when leaving it to default "Untitled 1")? To solve this you would need to escape characters like space before passing into `open` function.

**Review by PitNikola — CHANGES_REQUESTED** [review]
👆

**Comment by pernielsentikaer** [comment]
> Sorry, I just realised that I exported the icon with the background. Here's transparent one:
> 
> ray-so-icon
> 
> Now as for the issue that I had – I found the repro. Seems like `open` doesn't work when you have space in URL. This got broken on Big Sur when Apple decided to change its behavior for some reason. Could you please confirm that it happens whenever you have title with a space (e.g. when leaving it to default "Untitled 1")? To solve this you would need to escape characters like space before passing into `open` function.

It totally blended in with my background 😄 !

**Review by pernielsentikaer — APPROVED** [review]
I never had any problems with spaces in title, this is before latest commit:
QyITmArP8ZQ9vck7
2HvfjRZnie58T7hn

I retested it and the icon and the encoding works as well, nothing breaks, lets get it merged!

**Comment by Garrett1Tolbert** [comment]
> Seems like Apple fixed this issue in Monterey, I've just double checked, so it's only reproducible on Big Sur.
> 
> Validation for empty text area is still missing, but perhaps you could add it in the follow up PR @Garrett1Tolbert?
> 
> Merging it. Thanks for you contribution!

I completely missed that. I'll do that now

**Pull request merged**